azuredevops/provider.go

```diff
@@ -158,7 +547,12 @@ func providerConfigure(p *schema.Provider) schema.ConfigureContextFunc {
 			terraformVersion = "0.11+compatible"
 		}
 
-		client, err := client.GetAzdoClient(d.Get("personal_access_token").(string), d.Get("org_service_url").(string), terraformVersion)
+		token, err := GetAuthToken(ctx, d, AzIdentityFuncsReal{})
+		if err != nil {
+			return nil, diag.FromErr(err)
+		}
+
+		client, err := client.GetAzdoClient(token, d.Get("org_service_url").(string), terraformVersion)
```

Comment:
The call to `GetAzdoClient` will call `NewPatConnection`, what in turn will create a Basic Auth HTTP header using an OAuth token, what eventually will fail the authentication, because OAuth tokens must be passed as `Bearer`  Auth Header.

A even bigger issue is here that the Azure DevOps GO API uses a `struct` instead of an `interface` to pass a `Connection` around to the various REST API endpoint implementations and the `Connection` struct only supports a static auth header value. This is sufficient for a long living PAT but far from ideal for short living OAuth tokens. As long as this issue isn't fixed in the GO API is don't see a reliable way to support Service Principal Authentication for Azure DevOps.

@xuzhang3 any infos who's actually in charge for the GO API? Or is it abandoned again?

Reply:
That's a good point, I thought the OAuth tokens were valid long enough that it wouldn't practically be a concern. Looking at your other comment on the other thread I can test that and put it in a PR up to the Go SDK if someone is available to review it.

Reply:
@davidcorrigan714 The issue with the Azure DevOps GO SDK ist, that it's generated using a framework not created manually. So that's why most PRs got rejected there in past, because issues have to be fixed in the framework (which is not available here on GitHub) and not in the generated code in the repo you see. That the AzDO Team. Of course, the fact that the AzDO team maintains the GO SDK so carelessly makes things even worse. This is one of the main reasons why I personally stopped contributing to the project and to me I think Microsoft will focus on GitHub and slowly let AzDO die. That AzDO has much more stability issues compared to GitHub is a another good indicator that the manpower behind AzDO is very limited.

Reply:
Oh I'm well aware of the code-gen happening in all the SDKs. Pretty sure they completely lost or broke the Python SDK generator tooling a few years back, or lost the expertise that was running it. connection.go just doesn't show any history of ever being a generated file like the rest of the SDK is.

Don't want to derail this into GH vs AzDO too much. The reality is they don't have feature parity on the roadmap for the two platforms in certain areas and the rate they've been improving since trying to pitch GH as an AzDO alternative still puts it at least 5+ years out as a viable alternative for most of our use cases. We also do some "interesting" things with Pipelines that GH may never support in Actions.

Reply:
Any update? I would say ship a first version without refresh token support 😊

Reply:
The Azure DevOps team doesn't really have go developers, Azure DevOps is built in .NET, PowerShell and a bit of NodeJS and some web front-end stuff. The fact they even have a Go SDK amazes me, the team generates the Go and the Node SDK from the .NET API metadata as far as I know.

I don't think they maintain these libraries "carelessly", it's just that the person who wrote it, probably isn't available at the moment.

It's one of the big pains of building SDKs at all, you tend to need to build them in a plethora of languages, Python, .NET, JS, Go, Ruby etc, but the team that builds the SDK doesn't normally work with any of thee languages other than .NET. I see the same issue with many other frameworks and SDKs, the best SDK is written in for the language of the tool itself.

Reply:
The team maintaining TF az* providers is a central team. 
While you wait, consider [this workaround](https://github.com/geekzter/azure-identity-scripts/blob/123e9764195c8661e28a7dc189b2691447dad4bb/terraform/azure-devops/create-service-connection/terraform.tf#L27) that uses the `external` datasource to acquire a token with the Azure CLI:

```hcl
data external azdo_token {
  program                      = [
    "az", "account", "get-access-token", 
    "--resource", "499b84ac-1321-427f-aa17-267ca6975798", # Azure DevOps
    "--query","{accessToken:accessToken}",
    "-o","json"
  ]
}
provider azuredevops {
  org_service_url              = local.azdo_organization_url
  personal_access_token        = data.external.azdo_token.result.accessToken
}
```

Reply:
@jessehouwing yeah the choice of Go for Terraform always makes me a bit sad, it's the only reason I've picked up Go and I'm certainly not amazing at it though the Terraform stuff usually has plenty of examples to draw from.

Catching up on this thread again now and seeing where things are at with the code.